File: src/main/resources/project/server/checkNodeStatus.pl
#
#  Copyright 2015 Electric Cloud, Inc.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

=head1 NAME

checkNodeStatus - checks a node status

=head1 SYNOPSIS

=head1 DESCRIPTION

Checks a node status

=head1 LICENSE

Copyright (c) 2014 Electric Cloud, Inc.
All rights reserved.

=head1 AUTHOR

    ---

=head2 METHODS

=cut

# -------------------------------------------------------------------------
# Includes
# -------------------------------------------------------------------------
use ElectricCommander;
use ElectricCommander::PropMod qw(/myProject/modules);
use WebSphere::WebSphere;
use WebSphere::Util;

use warnings;
use strict;
$| = 1;

# -------------------------------------------------------------------------
# Variables
# -------------------------------------------------------------------------

my $configName        = trim(q($[configurationName]));
my $wsadminAbsPath    = trim(q($[wsadminAbsPath]));
my $nodeName          = trim(q($[nodeName]));


print "ConfigName: $configName\n";
print "WSadmin path $wsadminAbsPath\n";
print "Node Name: $nodeName\n";

# create args array
my @args  = ();
my %props = ();


#get an EC object
my $ec = new ElectricCommander();
$ec->abortOnError(0);

my $websphere = new WebSphere::WebSphere( $ec, $configName, $wsadminAbsPath );

my $file = 'check_node_status' . time() . '.py';
my $script = $ec->getProperty("/myProject/wsadmin_scripts/check_node_status.py")->getNodeText('//value');

open( my $fh, '>', $file ) or die "Cannot write to $file: $!";
print $fh $script;
close $fh;

my $shellcmd = $websphere->_create_runfile( $file, @args );
my $escapedCmdLine = $websphere->_mask_password($shellcmd);

print "WSAdmin command line:  $escapedCmdLine\n";
$props{'checkNodeStatusLine'} = $escapedCmdLine;
setProperties( $ec, \%props );

#execute command
print `$shellcmd 2>&1`;

#evaluates if exit was successful to mark it as a success or fail the step
if ( $? == SUCCESS ) {
    $ec->setProperty( "/myJobStep/outcome", 'success' );
}
else {
    $ec->setProperty( "/myJobStep/outcome", 'error' );
}

1;

File: src/main/resources/project/wsadmin_scripts/check_node_status.py
#
#  Copyright 2016 Electric Cloud, Inc.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#

import sys
myNode = r'''
$[nodeName]
'''.strip()

# This will be one of three: ALL_RUNNING, ALL_STOPPED, NODEAGENT_RUNNING
successCriteria = r'''
$[successCriteria]
'''.strip()

def showServerStatus(serverName, nodeName):
    serverObj = AdminControl.completeObjectName('WebSphere:type=Server,node=' + nodeName + ',name=' + serverName + ',*')
    if len(serverObj) > 0:
        serverStatus = AdminControl.getAttribute(serverObj, 'state')
    else:
        serverStatus = 'STOPPED'

    return serverStatus

def checkServerByCriteria(serverName, serverStatus, successCriteria):
    if successCriteria == 'ALL_RUNNING':
        if serverStatus != 'STARTED':
            return 0
    elif successCriteria == 'ALL_STOPPED':
        if serverStatus == 'STARTED':
            return 0
    else:
        if serverName == 'nodeagent' and serverStatus != 'STARTED':
            return 0
    # end of conditions
    return 1


nodeFound = 0
errorsCount = 0

for node in AdminConfig.list('Node').splitlines() :
    nodeName = AdminConfig.showAttribute(node, 'name')
    if myNode != nodeName :
        continue
    nodeFound = 1
    print '\n\n\n'
    for server in AdminConfig.list('Server', node).splitlines() :
        serverName = AdminConfig.showAttribute(server, 'name')
        serverStatus = showServerStatus(serverName, nodeName)
        # if serverStatus != 'STARTED':
        if not checkServerByCriteria(serverName, serverStatus, successCriteria):
            errorsCount += 1
        print '  Node: %s\nServer: %s\nStatus: %s\n' % (nodeName, serverName, serverStatus)
        print '==========\n'

if not nodeFound:
    print 'Node with name %s wasn\'t found' % myNode
    sys.exit(1)

if errorsCount > 0:
    print "Success Criteria wasn't met."
    sys.exit(1)

print "Success Criteria was met."
